keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,468 @@
+will queue another resync, and fall back to doing a full
+`iptables-restore` that time. (Note that there are no expected cases
+where a partial restore would fail but a full restore would succeed;
+this would always indicate a bug in the code.) Likewise, it always
+does a full resync when there are topology-related changes to Node
+labels, and it always does a full resync at least once every
+`iptablesSyncPeriod`.
+
+[the proposed PR]: https://github.com/kubernetes/kubernetes/pull/110268
+
+### Risks and Mitigations
+
+#### Catastrophic Failure
+
+In the event of bugs in the code, kube-proxy could fail
+catastrophically, entirely breaking the cluster. This would be "bad".
+
+#### Subtle Optimization Failure
+
+The code could also fail more subtly. In the worst case, every attempt
+to do a partial restore might fail, causing the code to fall back
+to the slower full `iptables-restore` on the next sync, with the end
+result that kube-proxy becomes _less_ efficient than the old code that
+would have just done a full restore the first time.
+
+Less obviously, the new code might reliably cause a fail-and-retry
+only in certain circumstances (eg, when a service goes from 1 endpoint
+to 0, or when a particular Service property changes).
+
+Since, as mentioned above, there are no _expected_ cases where a
+partial restore will fail, we could notice that this failure mode was
+happening by having a metric for how often a partial restore fails,
+and if we find that it is happening (in e2e tests or real clusters),
+we can then debug or revert the bad code.
+
+#### Subtle Synchronization Delays
+
+The new code needs to analyze Service/EndpointSlice changes to
+determine when the service-specific chains for each service need to be
+included in the `iptables-restore` input. If this analysis is
+incorrect, it might fail to update the iptables rules for a service in
+response to certain kinds of changes (eg, when a service goes from 1
+endpoint to 0, or when a particular Service property changes), causing
+the iptables rules to be stale until they eventually get fixed in a
+future full resync.
+
+In order to detect this, we would have to sanity-check every partial
+restore, to ensure that, if applied to the current kernel state, it
+would generate the same end result as the full restore we would
+otherwise have applied. Although we could do this with the
+`FakeIPTables` code in `pkg/util/iptables/testing`, that code is only
+intended for the sorts of small-ish test cases in the unit tests, and
+not at all optimized for handling the sort of very large rule sets
+that would exist in larger clusters (ie, the rule sets that would
+benefit the most from the partial restore feature).
+
+One possible approach for dealing with that would be to run the
+sanity-checking code only during the Alpha period, and only on
+"smaller" clusters where it could be done without too much performance
+loss. (If the code was going to fail in the way we are checking for
```

Comment:
I think this is something we might want to keep beyond alpha, behind an option, so we can toggle it in CI and use it to continue to ensure that we haven't introduced bugs

Reply:
There is no long-term reason to keep two implementations of this functionality and test them against each other in every CI run. We don't do that for any other feature in kubernetes, and there's no reason to do it here either. We're not suggesting this check because the code is especially likely to be buggy at all times; it's just that we want to make sure that we don't introduce bugs _right now_.

Reply:
> There is no long-term reason to keep two implementations of this functionality and test them against each other in every CI run. We don't do that for any other feature in kubernetes, and there's no reason to do it here either. We're not suggesting this check because the code is especially likely to be buggy at all times; it's just that we want to make sure that we don't introduce bugs right now.

Fair, but this only discussed _alpha_, presumably we'll still have two implementations until the feature is GA